pkg/gardenlet/controller/seed/seed_lease_control.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package seed
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	"github.com/gardener/gardener/pkg/logger"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+)
+
+func (c *Controller) seedLeaseAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.seedLeaseQueue.Add(key)
+}
+
+const leaseResyncSeconds = 2
+
+func (c *Controller) reconcileSeedLeaseKey(key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	seed, err := c.seedLister.Get(name)
+	if apierrors.IsNotFound(err) {
+		logger.Logger.Infof("[SEED LEASE] Stopping lease operations for Seed %s since it has been deleted", key)
+		c.seedLeaseQueue.Done(key)
+		return nil
+	}
+	if err != nil {
+		logger.Logger.Infof("[SEED LEASE] unable to retrieve Seed object %s from store: %v", key, err)
+		return err
+	}
+
+	if err := c.checkSeedConnection(seed); err != nil {
+		return fmt.Errorf("[SEED LEASE] cannot establish connection with Seed %s: %v", key, err)
+	}
+
+	condition := helper.GetCondition(seed.Status.Conditions, gardencorev1beta1.SeedGardenletReady)
+	if condition == nil {
+		condition, err = c.initGardenletReadyCondition(seed)
```

Comment:
Don't mutate cache objects. Deep copy them.

Why is it doing any mutation in the `reconcileSeedLeaseKey`? Any such operations should only be called in methods behind a workqueue.